src/contracts/PerformanceFee.sol

```diff
@@ -0,0 +1,159 @@
+    function _postWithdrawHook(uint256) internal virtual override {
+        lastTotalAssets = SafeCast.toUint128(_rawTotalAssets());
+    }
+
+    /// @dev Calculate the performance fee based on the increase in total assets
+    /// Needs to be called before any action that changes the liquidity provider shares. eg deposit and redeem
+    function _calcFee() internal {
+        uint256 newTotalAssets = _rawTotalAssets();
+
+        // Do not accrued a performance fee if the total assets has decreased
+        if (newTotalAssets <= lastTotalAssets) return;
+
+        uint256 assetIncrease = newTotalAssets - lastTotalAssets;
+        uint256 newFeesAccrued = (assetIncrease * fee) / FEE_SCALE;
+
+        // Save the new accrued fees back to storage
+        feesAccrued = SafeCast.toUint112(feesAccrued + newFeesAccrued);
+        // Save the new total assets back to storage
+        // This is be updated again in the post deposit and post withdraw hooks to include
+        // the assets deposited or withdrawn
+        lastTotalAssets = SafeCast.toUint128(newTotalAssets);
+
+        emit FeeCalculated(newFeesAccrued, assetIncrease);
+    }
+
+    function totalAssets() public view virtual override returns (uint256) {
+        uint256 totalAssetsBeforeFees = _rawTotalAssets();
+
+        // If the total assets have decreased, then we don't charge a performance fee
+        if (totalAssetsBeforeFees <= lastTotalAssets) return totalAssetsBeforeFees;
+
+        // Calculate the increase in assets since the last time fees were calculated
+        uint256 assetIncrease = totalAssetsBeforeFees - lastTotalAssets;
+
+        // Calculate the performance fee and remove from the total assets before new fees are removed
+        return totalAssetsBeforeFees - ((assetIncrease * fee) / FEE_SCALE);
+    }
+
+    /// @dev Calculate the total assets in the ARM, external withdrawal queue,
+    /// less liquidity assets reserved for the ARM's withdrawal queue and past accrued fees.
+    /// The accrued fees are from the last time fees were calculated.
+    function _rawTotalAssets() internal view returns (uint256) {
+        return super.totalAssets() - feesAccrued;
+    }
+
+    /// @notice Owner sets the performance fee on increased assets
+    /// @param _fee The performance fee measured in basis points (1/100th of a percent)
+    /// 10,000 = 100% performance fee
+    /// 500 = 5% performance fee
+    function setFee(uint256 _fee) external onlyOwner {
+        _setFee(_fee);
+    }
+
+    /// @notice Owner sets the account/contract that receives the performance fee
+    function setFeeCollector(address _feeCollector) external onlyOwner {
+        _setFeeCollector(_feeCollector);
+    }
+
+    function _setFee(uint256 _fee) internal {
+        require(_fee <= FEE_SCALE, "ARM: fee too high");
+
+        // Calculate fees up to this point using the old fee
+        _calcFee();
+
+        fee = SafeCast.toUint16(_fee);
```

Comment:
true, but I've defaulted to safety rather than saving some gas on liquidity provider functions.
The swap functions we want to be super gas efficient, the other functions are less important for gas

Reply:
It works for me 👍